cupyx/scipy/sparse/linalg/_solve.py

```diff
@@ -76,6 +78,287 @@ def lsqr(A, b):
+    allowed: m = n, m > n, or m < n. B is a vector of length m.
+    The matrix A may be dense or sparse (usually sparse).
+
+    Args:
+        A (ndarray, spmatrix or LinearOperator): The real or complex
+            matrix of the linear system. ``A`` must be
+            :class:`cupy.ndarray`, :class:`cupyx.scipy.sparse.spmatrix` or
+            :class:`cupyx.scipy.sparse.linalg.LinearOperator`.
+        b (cupy.ndarray): Right hand side of the linear system with shape
+            ``(m,)`` or ``(m, 1)``.
+        damp (float): Damping factor for regularized least-squares.
+            `lsmr` solves the regularized least-squares problem
+            ::
+
+                min ||(b) - (  A   )x||
+                    ||(0)   (damp*I) ||_2
+
+            where damp is a scalar. If damp is None or 0, the system
+            is solved without regularization.
+        atol, btol (float):
+            Stopping tolerances. `lsmr` continues iterations until a
+            certain backward error estimate is smaller than some quantity
+            depending on atol and btol.
+        conlim (float): `lsmr` terminates if an estimate of ``cond(A)`` i.e.
+            condition number of matrix exceeds `conlim`. If `conlim` is None,
+            the default value is 1e+8.
+        maxiter (int): Maximum number of iterations.
+
+    Returns:
+        tuple:
+            - `x` (ndarray): Least-square solution returned.
+            - `istop` (int): istop gives the reason for stopping
+
+                    0 means x=0 is a solution.
+
+                    1 means x is an approximate solution to A*x = B,
+                    according to atol and btol.
+
+                    2 means x approximately solves the least-squares problem
+                    according to atol.
+
+                    3 means COND(A) seems to be greater than CONLIM.
+
+                    4 is the same as 1 with atol = btol = eps (machine
+                    precision)
+
+                    5 is the same as 2 with atol = eps.
+
+                    6 is the same as 3 with CONLIM = 1/eps.
+
+                    7 means ITN reached maxiter before the other stopping
+                    conditions were satisfied.
+
+            - `itn` (int): Number of iterations used.
+            - `normr` (float): ``norm(b-Ax)``
+            - `normar` (float): ``norm(A^T (b - Ax))``
+            - `norma` (float): ``norm(A)``
+            - `conda` (float): Condition number of A.
+            - `normx` (float): ``norm(x)``
+
+    .. seealso:: :func:`scipy.sparse.linalg.lsmr`
+
+    References:
+        D. C.-L. Fong and M. A. Saunders, "LSMR: An iterative algorithm for
+        sparse least-squares problems", SIAM J. Sci. Comput.,
+        vol. 33, pp. 2950-2971, 2011.
+    """
+    A = _interface.aslinearoperator(A)
+    b = b.squeeze()
+
+    m, n = A.shape
+    minDim = min([m, n])
+
+    if maxiter is None:
+        maxiter = minDim
+
+    u = b
```

Comment:
This breaks the contents of `b` during the iterations. Would you assign a copy of `b` here?

Reply:
b was only used to get beta only single time and after that, there is no use of b. So I think this works well.

Reply:
Now `u` refers to the same object with `b`, so when the contents of `u` is changed, that of `b` is also.

Reply:
The contents of `u` is changed here: https://github.com/cupy/cupy/blob/405fe1a67a67b37e0d62acaaa84ca18b1a4e5332/cupyx/scipy/sparse/linalg/_solve.py#L170

Reply:
I think as there is no use of b(except for beta which is calculated before modifying u) so making a copy of it would be needless.

Reply:
`b` and `n` share the same ndarray, so modifying `u` also changes the contents of `b`. Please check if the following assertion passes.

```Python
import numpy as np
import scipy.sparse
import cupy as cp
from cupy import testing
from cupyx.scipy.sparse import csr_matrix
from cupyx.scipy.sparse.linalg import lsmr

m = 1000
a = scipy.sparse.random(m, m, density=0.001)
b = np.random.rand(m)
a = csr_matrix(a)
b = cp.array(b)
b0 = b.copy()
lsmr(a,b)
testing.assert_array_equal(b, b0)
print('pass')
```

Reply:
Ok, now I got it.